### ISSUE TITLE
Implement gradient checkpointing for larger models (Klein 9B, Dev)

### ISSUE THREAD

**Issue by VincentGourbin** [VincentGourbin/flux-2-swift-mlx#38]
## Problem

Training LoRA on larger models (Klein 9B, Dev) with DOP (Differential Output Preservation) causes OOM on 96GB M3 Max due to the additional memory required for:
- Extra forward passes for prior preservation
- 6 additional text embeddings held in memory
- Full backward pass without memory optimization

## Current Status

| Model | DOP | Gradient Checkpointing | Status |
|-------|-----|------------------------|--------|
| Klein 4B | ✅ | N/A | ✅ Works |
| Klein 9B | ❌ | Not implemented | ✅ Works without DOP |
| Klein 9B | ✅ | Not implemented | ❌ OOM |
| Dev | ✅ | Not implemented | ❌ OOM (expected) |

### Attempted Solution (Failed)

We tried wrapping the forward pass with MLX's `checkpoint()` function:

```swift
let checkpointedForward = checkpoint { inputs -> [MLXArray] in
    model(hiddenStates: inputs[0], ...)
}
```

**Result**: Gradients were always 0. The model is captured by reference, not included in the compute graph. MLX `checkpoint()` expects a pure function `[MLXArray] -> [MLXArray]`.

### Correct Solution

**Layer-wise checkpointing**: Modify `Flux2Transformer2DModel` to wrap each transformer block individually with checkpoint. This is more complex as it requires modifying the model architecture itself.

Reference implementation approach:
```swift
// In Flux2Transformer2DModel
for block in transformerBlocks {
    // Wrap each block's forward pass with checkpoint
    hiddenStates = checkpoint { inputs in
        block.forward(inputs[0], inputs[1], ...)
    }([hiddenStates, encoderHiddenStates, ...])
}
```

## Files to Modify

- `Sources/Flux2Core/Models/Transformer/Flux2Transformer2DModel.swift`
- `Sources/Flux2Core/Models/Transformer/Flux2TransformerBlock.swift`
- `Sources/Flux2Core/Training/Loop/SimpleLoRATrainer.swift`

## References

- MLX checkpoint documentation: [MLX Compilation](https://ml-explore.github.io/mlx/build/html/python/transforms.html#mlx.core.checkpoint)
- Similar implementation in Python: [ai-toolkit](https://github.com/ostris/ai-toolkit)
- MLX Swift examples: [mlx-swift-examples](https://github.com/ml-explore/mlx-swift-examples)

## Help Wanted

Any contributions to implement proper layer-wise gradient checkpointing are welcome! This would enable:
- Training Klein 9B with DOP
- Training Dev model
- Generally lower memory footprint for all training configurations

If you have experience with MLX gradient checkpointing or similar memory optimization techniques, please reach out or submit a PR.

**Comment by VincentGourbin** [comment]
Resolved by PR #42. Layer-wise gradient checkpointing implemented with explicit param passing for LoRA gradient flow. Tested: Klein 4B+DOP (42GB peak, 30% reduction), Klein 9B+DOP (94.7GB peak, 130 steps, no OOM on 96GB). Enable via `gradient_checkpointing: true` in YAML. Trade-off: ~1.6x slower step time.
